dill/session.py

```diff
@@ -0,0 +1,220 @@
+    if exclude_extra is not None:
+        filters.update([(EXCLUDE, exclude_extra)])
+    if include_extra is not None:
+        filters.update([(INCLUDE, include_extra)])
+
+    namespace = filters.filter_namespace(vars(main), obj=obj)
+    if namespace is vars(main):
+        return main
+
+    main = ModuleType(main.__name__)
+    vars(main).update(namespace)
+    return main
+
+def dump_session(filename: Union[PathLike, BytesIO] = '/tmp/session.pkl',
+                 main: Union[str, ModuleType] = '__main__',
+                 byref: bool = False,
+                 exclude: Union[Filter, Iterable[Filter]] = None,
+                 include: Union[Filter, Iterable[Filter]] = None,
+                 **kwds) -> NoReturn:
+    """pickle the current state of __main__ to a file"""
+    protocol = settings.protocol
+    if isinstance(main, str):
+        main = _import_module(main)
+    original_main = main
+    if byref:
+        #NOTE: *must* run before _filter_objects()
+        main = _stash_modules(main)
+    main = _filter_objects(main, exclude, include, obj=original_main)
+    if hasattr(filename, 'write'):
+        f = filename
+    else:
+        f = open(filename, 'wb')
+    try:
+        pickler = Pickler(f, protocol, **kwds)
+        pickler._main = main     #FIXME: dill.settings are disabled
+        pickler._byref = False   # disable pickling by name reference
+        pickler._recurse = False # disable pickling recursion for globals
+        pickler._session = True  # is best indicator of when pickling a session
+        pickler._first_pass = True
+        if main is not original_main:
+            pickler._original_main = original_main
+        pickler.dump(main)
+    finally:
+        if f is not filename:  # If newly opened file
+            f.close()
+    return
+
+def load_session(filename: Union[PathLike, BytesIO] = '/tmp/session.pkl',
+                 main: ModuleType = None,
+                 **kwds) -> NoReturn:
+    """update the __main__ module with the state from the session file"""
+    if main is None:
+        main = _main_module
+    if hasattr(filename, 'read'):
+        f = filename
+    else:
+        f = open(filename, 'rb')
+    try: #FIXME: dill.settings are disabled
+        unpickler = Unpickler(f, **kwds)
+        unpickler._main = main
+        unpickler._session = True
+        module = unpickler.load()
+        unpickler._session = False
+        main.__dict__.update(module.__dict__)
+        _restore_modules(unpickler, main)
+    finally:
+        if f is not filename:  # If newly opened file
+            f.close()
+    return
+
+#############
+#  IPython  #
+#############
+
+def ipython_filter(*, keep_input=True, keep_output=False):
+    """filter factory for IPython sessions (can't be added to settings currently)
+
+    Usage:
+    >>> from dill.session import *
+    >>> dump_session(exclude=[ipython_filter()])
+    """
+    if not __builtins__.get('__IPYTHON__'):
```

Comment:
Note that this is checked already in `dill._dill`. Probably fine as is.

Reply:
Yup. But it's not stored there and a try/except block seemed overkill here...